src/promptflow/promptflow/executor/_line_execution_process_pool.py

```diff
@@ -411,6 +404,55 @@ def _generate_thread_status_messages(self, pool: ThreadPool, total_count: int):
             msgs.append("Processing Lines: " + ", ".join(lines) + ".")
         return msgs
 
+    def _determine_worker_count(self):
+        worker_count = get_int_env_var("PF_WORKER_COUNT")
+
+        # Starting a new process in non-fork mode requires to allocate memory. Calculate the maximum number of processes
+        # based on available memory to avoid memory bursting.
+        estimated_available_worker_count = get_available_max_worker_count() if not self._use_fork else None
+
+        # If the environment variable PF_WORKER_COUNT exists and valid, use the value as the worker_count.
+        if worker_count is not None and worker_count > 0:
+            self._log_set_worker_count(worker_count, estimated_available_worker_count)
+            return worker_count
+
+        return self._calculate_worker_count(estimated_available_worker_count)
+
+    def _log_set_worker_count(self, worker_count, estimated_available_worker_count):
+        bulk_logger.info(
+            f"Process count set to {worker_count} based on the 'PF_WORKER_COUNT' environment variable.")
+        if estimated_available_worker_count is not None and estimated_available_worker_count < worker_count:
+            bulk_logger.warning(
+                f"The estimated available worker count calculated based on the system available memory "
+                f"is {estimated_available_worker_count}, but the PF_WORKER_COUNT is set to "
+                f"{worker_count}. This may affect optimal memory usage and performance.")
+
+    def _calculate_worker_count(self, estimated_available_worker_count):
```

Comment:
I think this function looks unnecessary, it would be tricky to ask "What if estimated_available_worker_count is None"?
Then the answer would be "It is always not None when use_fork is false because we have such logic outside this function."
It is not a good practice to add such assumption for a function.

Maybe the code could be moved to `determine_worker_count`

Reply:
fixed